plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php

```diff
@@ -0,0 +1,209 @@
+					&& false === strpos( $menu_item[0], $badge ) ) {
+
+					$menu[ $index ][0] .= $badge; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+
+					// One menu item with a badge is more than enough.
+					break;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Preload settings to make them available to the Payments settings page frontend logic.
+	 *
+	 * Added keys will be available in the window.wcSettings.admin object.
+	 *
+	 * @param array $settings The settings array.
+	 *
+	 * @return array Settings array with additional settings added.
+	 */
+	public function preload_settings( array $settings ): array {
+		// We only preload settings in the WP admin.
+		if ( ! is_admin() ) {
+			return $settings;
+		}
+
+		// No preloaded settings for now.
+
+		return $settings;
+	}
+
+	/**
+	 * Adds promo note IDs to the list of allowed ones.
+	 *
+	 * @param array $promo_notes Allowed promo note IDs.
+	 *
+	 * @return array The updated list of allowed promo note IDs.
+	 */
+	public function add_allowed_promo_notes( array $promo_notes = array() ): array {
+		try {
+			$providers = $this->payments->get_payment_providers( WC()->countries->get_base_country() );
+		} catch ( Exception $e ) {
+			// In case of an error, bail.
+			return $promo_notes;
+		}
+
+		// Add all incentive promo IDs to the allowed promo notes list.
+		foreach ( $providers as $provider ) {
+			if ( ! empty( $provider['_incentive']['promo_id'] ) ) {
+				$promo_notes[] = $provider['_incentive']['promo_id'];
+			}
+		}
+
+		return $promo_notes;
+	}
+
+	/**
+	 * Check if the store has any enabled gateways (including offline payment methods).
+	 *
+	 * @return bool True if the store has any enabled gateways, false otherwise.
+	 */
+	private function store_has_enabled_gateways(): bool {
+		$gateways         = WC()->payment_gateways->get_available_payment_gateways();
+		$enabled_gateways = array_filter(
+			$gateways,
+			function ( $gateway ) {
+				return 'yes' === $gateway->enabled;
+			}
+		);
+
+		return ! empty( $enabled_gateways );
+	}
+
+	/**
+	 * Check if the store has any payment providers that need an action/attention.
+	 *
+	 * This includes gateways that are enabled but not configured (need setup).
+	 *
+	 * @return bool True if the store has enabled gateways that need attention, false otherwise.
+	 */
+	private function store_has_providers_with_action(): bool {
+		try {
+			$providers = $this->payments->get_payment_providers( WC()->countries->get_base_country() );
+		} catch ( Exception $e ) {
+			// In case of an error, just return false.
+			return false;
+		}
+
+		// Go through the providers and check if any of them need attention from the user.
+		foreach ( $providers as $provider ) {
+			// Handle payment gateways and offline payment methods that need setup.
+			if (
+				in_array( $provider['_type'], array( Payments::PROVIDER_TYPE_GATEWAY, Payments::PROVIDER_TYPE_OFFLINE_PM ), true ) &&
+				! empty( $provider['state']['enabled'] ) &&
```

Comment:
I think that if it needs setup, then it won't be enabled yet. at least that's what I noticed during my testing with Stripe, WooPayments.

Reply:
> I think that if it needs setup, then it won't be enabled yet.

This is very much at each PE's discretion. Yes, WooPayments and the Stripe official PE work like that, but not all of them.

> in this case, we'll show the badge even if it's enabled, but needs setup

Quite the contrary. If it is NOT enabled but needs setup, we won't show the badge (the conditions have AND between them). But you do raise a good issue. We should not look at `enabled`, just at `needs_setup`. Changing it.